src/main/java/bio/terra/cli/model/GlobalContext.java

```diff
@@ -0,0 +1,137 @@
+package bio.terra.cli.model;
+
+import bio.terra.cli.utils.FileUtils;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This POJO class represents an instance of the Terra CLI global context. This is intended
+ * primarily for authentication and connection-related context values that will span multiple
+ * workspaces.
+ */
+public class GlobalContext {
+  private static final Logger logger = LoggerFactory.getLogger(GlobalContext.class);
+
+  // global auth context =  list of Terra users, CLI-generated key of current Terra user
+  public Map<String, TerraUser> terraUsers;
+  public String currentTerraUserKey;
+
+  // global server context = service uris, environment name
+  public ServerSpecification server;
+
+  private GlobalContext() {
+    this.terraUsers = new HashMap<>();
+    this.server = null;
+  }
+
+  // ====================================================
+  // Persisting on disk
+
+  /**
+   * Read in an instance of this class from a JSON-formatted file in the global context directory.
+   * If there is no existing file, this method returns an object populated with default values.
+   *
+   * @return an instance of this class
+   */
+  public static GlobalContext readFromFile() {
+    // try to read in an instance of the global context file
+    GlobalContext globalContext = null;
+    try {
+      globalContext =
+          FileUtils.readOutputFileIntoJavaObject(
+              resolveGlobalContextFile().toFile(), GlobalContext.class);
+    } catch (IOException ioEx) {
+      logger.error("Error reading in global context file.", ioEx);
+    }
+
+    // if the global context file does not exist, return an object populated with default values
+    if (globalContext == null)
+      try {
+        globalContext = new GlobalContext();
+        globalContext.server = ServerSpecification.fromJSONFile(DEFAULT_SERVER_FILENAME);
+      } catch (IOException ioEx) {
+        logger.error("Error reading in default server file. ({})", DEFAULT_SERVER_FILENAME, ioEx);
+      }
+
+    return globalContext;
+  }
+
+  /** Write an instance of this class to a JSON-formatted file in the global context directory. */
+  public void writeToFile() {
+    try {
+      FileUtils.writeJavaObjectToFile(resolveGlobalContextFile().toFile(), this);
+    } catch (IOException ioEx) {
+      logger.error("Error persisting global context.", ioEx);
+    }
+  }
+
+  // ====================================================
+  // Auth
+
+  /** Getter for the current Terra user. Returns null if no current user is defined. */
+  public TerraUser getCurrentTerraUser() {
```

Comment:
Changed for now. I'm not sure this makes it much clearer and it just added lines of code. But perhaps this will be an easier decision as more classes start to access the globalcontext ojbect.

Reply:
ack. I'm not an Optional zealot. Sometimes it's great, sometimes not. And often the difference is how all-in a codebase is on Optional and the fancy usage-side utility methods. Jay may have some ideas to contribute here.